dwave/optimization/src/nodes/indexing.cpp

```diff
@@ -1596,6 +1597,156 @@ std::span<const ssize_t> BasicIndexingNode::shape(const State& state) const {
+    std::copy(shape0.begin(), shape0.end(), std::back_inserter(shape));
+
+    // On the concatenation axis we sum the axis dimension sizes
+    for (auto it = std::next(array_ptrs.begin()); it != array_ptrs.end(); it++) {
+       shape[axis] = shape[axis] + (*it)->shape()[axis];
+    }
+
+    return shape;
+}
+
+ConcatenateNode::ConcatenateNode(std::vector<ArrayNode*> array_ptrs, const ssize_t axis)
+        : ArrayOutputMixin(make_concatenate_shape(array_ptrs, axis)) {
+
+    axis_ = axis;
+    array_ptrs_ = array_ptrs;
+
+    for (auto it = array_ptrs.begin(); it != array_ptrs.end(); it++) {
+        this->add_predecessor((*it));
+    }
+}
+
+double const* ConcatenateNode::buff(const State& state) const {
+    return data_ptr<ArrayNodeStateData>(state)->buff();
+}
+
+std::span<const Update> ConcatenateNode::diff(const State& state) const {
+    return {};
+}
+
+void ConcatenateNode::initialize_state(State& state) const {
+    int index = topological_index();
+    assert(index >= 0 && "must be topologically sorted");
+    assert(static_cast<int>(state.size()) > index && "unexpected state length");
+    assert(state[index] == nullptr && "already initialized state");
+
+    std::vector<double> values;
+    values.reserve(size());
+
+    // Prefix is defined over dimensions 0..axis-1
+    std::vector<std::vector<ssize_t>> prefix_dims(axis_);
+    for (ssize_t dim = 0; dim < prefix_dims.size(); dim++) {
+        for (ssize_t i = 0; i < shape()[dim]; i++) {
+            prefix_dims[dim].push_back(i);
+        }
+    }
+
+    // Suffix is defined over dimensions axis+1..ndim
+    std::vector<std::vector<ssize_t>> suffix_dims(ndim() - axis_ - 1);
+    for (ssize_t dim = 0; dim < suffix_dims.size(); dim++) {
+        for (ssize_t i = 0; i < shape()[axis_ + 1 + dim]; i++) {
+            suffix_dims[dim].push_back(i);
+        }
+    }
+
+    auto prefix_prod = cartesian_product(prefix_dims);
+    auto suffix_prod = cartesian_product(suffix_dims);
+
+    for (auto prefix : prefix_prod) {
+        for (ssize_t arr_i = 0; arr_i < array_ptrs_.size(); arr_i++) {
+            size_t arr_axis = array_ptrs_[arr_i]->shape()[axis_];
+            for (ssize_t arr_axis_i = 0; arr_axis_i < arr_axis; arr_axis_i++) {
+                for (auto suffix : suffix_prod) {
+                    std::vector<ssize_t> indices;
+                    indices.insert(indices.begin(), prefix.begin(), prefix.end());
+                    indices.insert(indices.begin() + prefix_dims.size(), arr_axis_i);
+                    indices.insert(indices.begin() + prefix_dims.size() + 1, suffix.begin(), suffix.end());
+
+                    ssize_t idx = ravel_multi_index(array_ptrs_[arr_i]->strides(), indices);
+                    values.emplace_back(array_ptrs_[arr_i]->buff(state)[idx]);
+                }
+            }
+        }
+    }
+
+    state[index] = std::make_unique<ArrayNodeStateData>(std::move(values));
+}
+
+void ConcatenateNode::commit(State& state) const {
+
```

Comment:
I think this can just be `data_ptr<ArrayNodeStateData>(state)->commit();` That's one of the benefits to using `ArrayNodeStateData`, it handles committing/reverting.